src/plugin.ts

```diff
@@ -63,12 +63,22 @@ export default {
 
 		// we need to share some logic between the plugin and the routeGuard
 		// maybe there is a better way to do it
-		routeGuardInternal.value = async () => {
-			if (!sessionToken.value && isFetchSessionWasNeverCalled.value) {
-				await fetchSession();
-			}
-			return !!unref(sessionToken);
-		};
+		routeGuardInternal.value = () =>
+			new Promise((resolve, reject) => {
+				if (!sessionToken.value && isFetchSessionWasNeverCalled.value) {
+					fetchSession().catch(reject);
+				}
+
+				// if the session is loading we want to wait for it to finish before resolving
+				if (isSessionLoading.value) {
+					watch(
+						() => isSessionLoading.value,
+						() => !isSessionLoading.value && resolve(!!unref(sessionToken))
+					);
+				} else {
+					resolve(!!unref(sessionToken));
+				}
+			});
```

Comment:
I didn't test it, but we may be able to use [eager watcher](https://vuejs.org/guide/essentials/watchers.html#eager-watchers) to avoid this condition

```suggestion
				watch(
						() => isSessionLoading.value,
						() => !isSessionLoading.value && resolve(!!unref(sessionToken)),
						{ immediate: true }
					);
			});
```

Reply:
Thanks!
I wasn't familiar with this option